server/src/main/java/org/elasticsearch/tasks/CancellableTasksTracker.java

```diff
@@ -31,21 +31,45 @@ public CancellableTasksTracker(T[] empty) {
     }
 
     private final Map<Long, T> byTaskId = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
-    private final Map<TaskId, T[]> byParentTaskId = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+    private final Map<TaskId, Map<Long, T[]>> byParentTaskId = ConcurrentCollections.newConcurrentMapWithAggressiveConcurrency();
+
+    /**
+     * Gets the cancellable children of a parent task.
+     *
+     * Note: children of non-positive request IDs (e.g., -1) may be grouped together.
+     */
+    public Stream<T> getChildrenByRequestId(TaskId parentTaskId, long childRequestId) {
```

Comment:
I wonder if doing a linear search using the previous approach from `Map<TaskId, T[]>` would be enough and will simplify all this code. For that we would need to store the `requestId` in the `CancellableTaskHolder` or something like that. 
I mention this since the `put` method mentions that it's expected to be called only once per task and now that we have this notion of `requestId` that you mention that it can be -1 multiple times (when is that possible? we set the requestId for all incoming requests and local requests are not tracked here)